x-pack/plugins/fleet/server/services/artifacts/artifacts.ts

```diff
@@ -82,6 +84,50 @@ export const createArtifact = async (
   return esSearchHitToArtifact({ _id: id, _source: newArtifactData });
 };
 
+export const bulkCreateArtifacts = async (
+  esClient: ElasticsearchClient,
+  artifacts: NewArtifact[],
+  refresh = false
+): Promise<{ artifacts?: Artifact[]; errors?: Error[] }> => {
+  const ids = artifacts.map((artifact) => uniqueIdFromArtifact(artifact));
+  const newArtifactsData = artifacts.map((artifact) =>
+    newArtifactToElasticsearchProperties(artifact)
+  );
+
+  const body = ids
+    .map((id, index) => [
+      {
+        create: {
+          _id: id,
+        },
+      },
+      newArtifactsData[index],
+    ])
+    .flatMap((arr) => arr);
+
+  const res = await withPackageSpan('Bulk create fleet artifacts', () =>
+    esClient.bulk({
+      index: FLEET_SERVER_ARTIFACTS_INDEX,
+      body,
+      refresh,
+    })
+  );
+  if (res.errors) {
+    const nonConflictErrors = res.items
+      .filter((item) => item.create?.status !== 409)
+      .map((item) => new Error(item.create?.error?.reason));
```

Comment:
Here too you can iterate once through the `items` instead of twice. Something like 

```json5
const nonConflictErrors = res.items.reduce<Error[]>((acc, item) => {
      if (item.create?.status !== 409) {
        acc.push(new Error(item.create?.error?.reason));
      }
      return acc;
    }, []);
```

Reply:
I personally prefer the functional style.